File: InheriTypes.cpp
/*Demonstrate different types of inheritance in C++.*/
#include<iostream>
using namespace std;
class A
{
    public:
        void a()
        {
            cout<<"I'm class A"<<endl;
        }
};
class B
{
    public:
        void b()
        {
            cout<<"I'm class B"<<endl;
        }
};

// Simple Inheritance
class C : public A
{
    public:
    void c()
    {
        cout<<"I'm class C"<<endl;
    } 
};

// Multilevel Inheritance
class D : public A
{
    public:
    void d()
    {
        cout<<"I'm class D"<<endl;
    }
};
class E : public D
{
    public:
    void e()
    {
        cout<<"I'm class E"<<endl;
    }
};

// Multiple Inheritance
class F : public A, public B
{
    public:
    void f()
    {
        cout<<"I'm class F"<<endl;
    }
};

// Hierarchial Inheritance
class G : public A
{
    public:
    void g()
    {
        cout<<"I'm class G"<<endl;
    }
};
class H : public A
{
    public:
    void h()
    {
        cout<<"I'm class H"<<endl;
    }
};

// Hybrid Inheritance
class I : public C, public B
{
    public:
    void i()
    {
        cout<<"I'm class I"<<endl;
    } 
};
int main()
{
    // Simple Inheritance
    cout<<"Simple Inheritance"<<endl;
    C obj1;
    obj1.a();
    obj1.c();

    // Multilevel Inheritance
    cout<<"Multilevel Inheritance"<<endl;
    E obj2;
    obj2.a();
    obj2.d();
    obj2.e();

    // Multiple Inheritance
    cout<<"Multiple Inheritance"<<endl;
    F obj3;
    obj3.a();
    obj3.b();
    obj3.f();

    // Hierarchial Inheritance
    cout<<"Hierarchial Inheritance"<<endl;
    G obj4;
    obj4.a();
    obj4.g();

    H obj5;
    obj5.a();
    obj5.h();

    // Hybrid Inheritance
    cout<<"Hybrid Inheritance"<<endl;
    I obj6;
    obj6.a();
    obj6.c();
    obj6.i();
    obj6.b();

    return 0;
}

File: InheriTypes.java
class A
{
    public void a()
    {
        System.out.println("I'm class A");
    }
}
class B
{
    public void b()
    {
        System.out.println("I'm class B");
    }
}

// Simple Inheritance
class C extends A
{
    public void c()
    {
        System.out.println("I'm class C");
    } 
}

// Multilevel Inheritance
class D extends A
{
    public void d()
    {
        System.out.println("I'm class D");
    } 
}
class E extends D
{
    public void e()
    {
        System.out.println("I'm class E");
    } 
}

// Multiple Inheritance
// Java doesn't support Multiple Inheritance

// Hierarchial Inheritance
class F extends A
{
    public void f()
    {
        System.out.println("I'm class F");
    } 
}
class G extends A
{
    public void g()
    {
        System.out.println("I'm class G");
    } 
}

// Hybrid Inheritance
// Java doesn't support Hybrid Inheritance

public class InheriTypes
{
    public static void main(String[] args)
    {
        // Simple Inheritance
        System.out.println("Simple Inheritance");
        C obj1=new C();
        obj1.a();
        obj1.c();

        // Multilevel Inheritance
        System.out.println("Multilevel Inheritance");
        E obj2=new E();
        obj2.a();
        obj2.d();
        obj2.e();

        // Hierarchial Inheritance
        System.out.println("Hierarchial Inheritance");
        F obj3=new F();
        obj3.a();
        obj3.f();

        G obj4=new G();
        obj4.a();
        obj4.g();
    }
}